### DISCUSSION TITLE
HDRLoader: Handle RGBA source Buffer

### DISCUSSION THREAD

**Discussion by ThomasGorisse** [google/filament#4610] [Ideas]
Hi,

`HDRLoader` is doing a great job coupled to the `SpecularFilter`.
https://github.com/google/filament/blob/9ff195ce19d621db19eea925c064558b1fd2c2b9/android/filament-utils-android/src/main/cpp/HDRLoader.cpp#L36-L91
But I have a case where my source buffer is in RGBA16F (ARCore estimated cubemap).
This involve that I have to create a new buffer in order to remove the alpha channel.
This wouldn't be a big issue in a normal usage case but in mine, I have to do it on every frame.

Would it possible to add an input format param to `nCreateHDRTexture`?

If it's not applicable do you have any idea of the best way/less consuming way to strip the alpha channel from the buffer ?

Thanks

**Comment by ThomasGorisse** [comment]
My alpha bytes skipping solution doesn't seems to work:
```kotlin
HDRLoader.createEnvironment(ByteBuffer.allocateDirect(
                arImages[0].width * arImages[0].height *
                        arImages.size *
                        // RGB Bytes per pixel
                        6 * 2
            ).apply {
                // Use the device hardware's native byte order
                order(ByteOrder.nativeOrder())
                val rgbaBytes = ByteArray(8) // ARGB Bytes per pixel
                arImages.forEach { image ->
                    image.planes[0].buffer.let { imageBuffer ->
                        while (imageBuffer.hasRemaining()) {
                            // Only take the RGB channels
                            put(rgbaBytes.apply {
                                imageBuffer.get(this)
                            } // Skip the Alpha channel
                                .sliceArray(0..5))
                        }
                    }
                }
            })
```

**Comment by ThomasGorisse** [comment]
Sorry,

I just discovered what an hdr file is and that it's not only a RGB container at all.
My mind just went from ARCore gives me 6 HDR images and filament can create a texture from an HDR file to let's create an HDR texture from the image content.

**Reply by romainguy** [reply]
Why do you want to skip the alpha channel? RGB16F is not a native format and drivers/GPUs (or in the case of Vulkan/Filament) have to expand it to RGBA16F anyway.

**Reply by ThomasGorisse** [reply]
Because of this on the AR Core part:
[An array that contains 6 images in AIMAGE_FORMAT_RGBA_FP16 format for the HDR cubemap. The memory layout for the image data is identical to GL_RGBA16F](https://developers.google.com/ar/reference/java/com/google/ar/core/LightEstimate#public-methods_1)

And this on the Filament part:
https://github.com/google/filament/blob/9ff195ce19d621db19eea925c064558b1fd2c2b9/android/filament-android/src/main/java/com/google/android/filament/Texture.java#L1097
